static/js/protein/data_processing_utils.ts

```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
     return proteinDescription;
   }
 }
+
+/**
+ * Given array, key-maker function, and value-maker function, construct object from array with string-valued keys.
+ * This is essentially a TS implementation of a python dictionary comprehension.
+ *
+ * Unfortunately generic index signature parameters aren't supported by TS yet, so we enforce that keys are strings
+ * (since they're automatically converted to strings anyway).
+ *
+ * References:
+ *  - https://linuxtut.com/en/87ab3f313a75b547c278/
+ *  - https://stackoverflow.com/questions/13315131/enforcing-the-type-of-the-indexed-members-of-a-typescript-object
+ *
+ */
+export function objectFromArray<T, V>(
+  arr: T[],
+  keyFunc: (arrElt: T) => string,
+  valFunc: (arrElt: T) => V
+): { [key: string]: V } {
+  return arr.reduce(
+    (obj, elt) => Object.assign(obj, { [keyFunc(elt)]: valFunc(elt) }),
+    {}
+  );
+}
+
+/**
+ * Given an (unpacked) array of arrays, return its transpose (after trimming arrays to length of shortest one).
+ * This is essentially a TS implementation of python's built-in zip function.
+ *
+ * Adapted from implementation by Matt Mancini
+ *
+ * Thanks also to Carlos Israel Ortiz Garcia and Jack Davison for the below references:
+ *  - https://ncjamieson.com/typing-zipwith/#typing-zip
+ *  - https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types
+ */
+// Recursive elementwise type inference on array of arrays
+type Zip<T> = T extends [Array<infer U>, ...infer Z] ? [U, ...Zip<Z>] : [];
+export function zip<T extends unknown[][]>(...arrays: T): Array<Zip<T>> {
+  const minLength = Math.min(...arrays.map((arr) => arr.length));
+  return Array.from(Array(minLength).keys()).map((_, i) => {
+    return arrays.map((array) => array[i]) as Zip<T>;
+  });
+}
+
+/**
+ * Convert DCID of form bio/<id> to id.
+ * Note: probably should be doing dcid: `bio/${string}`, but currently causes many type errors
+ */
+export function idFromDCID(dcid: string): string {
```

Comment:
Since there are so many ID types floating around in this module -- perhaps specify that this is the `proteinId`? I see places here where this sometimes include species -- specificity will go a long way towards reducing errors.

Also - only export functions that will be used outside the module (applies to everything else in here).

Reply:
Oh ok thanks for pointing this out - the only two ID types for the data of the interaction graph which are true IDs (i.e. used to ID nodes/links in the graph) are protein ID and interaction ID, and `idFromDCID` and `dcidFromID` apply to both.  However in `chart.ts` there's stuff like `elementID`, `barIDFunc`, `circleIDFunc`.  Perhaps I could make `idFromDCID` and `dcidFromID` static methods in the new interaction graph component?

Also I think I wrote species ID in a comment somewhere without realizing it was confusing - I'll just revise the comments so there's no mention of IDs in the graph part other than protein ID and interaction ID.

Reply:
I renamed these functions to `ppiIDFromDCID` and `ppiDCIDFromID` and replaced all mentions of `species ID` with `species name`.  

To make this a bit safer I also added an assertion in `ppiIDFromDCID` to enforce that the input cannot start with `"bio/"`, and a new type `bioDCID = `bio/${string}` to enforce that all variables labeled DCID throughout the code start with `"bio/"`.  (please let me know if you think this is overkill I can revert back to just string type for everything).